openapi_python_client/parser/properties/union.py

```diff
@@ -189,3 +220,98 @@ def validate_location(self, location: oai.ParameterLocation) -> ParseError | Non
+            discriminators.extend(new_discriminators)
+        else:
+            flattened.append(sub_prop)
+    return flattened, discriminators
+
+
+def _parse_discriminator(
+    data: oai.Discriminator,
+    subtypes: list[PropertyProtocol],
+    schemas: Schemas,
+) -> DiscriminatorDefinition | PropertyError:
+    from .model_property import ModelProperty
+
+    # See: https://spec.openapis.org/oas/v3.1.0.html#discriminator-object
+
+    # Conditions that must be true when there is a discriminator:
+    # 1. Every type in the anyOf/oneOf list must be a $ref to a named schema, such as
+    #    #/components/schemas/X, rather than an inline schema. This is important because
+    #    we may need to use the schema's simple name (X).
+    # 2. There must be a propertyName, representing a property that exists in every
+    #    schema in that list (although we can't currently enforce the latter condition,
+    #    because those properties haven't been parsed yet at this point.)
+    #
+    # There *may* also be a mapping of lookup values (the possible values of the property)
+    # to schemas. Schemas can be referenced either by a full path or a name:
+    #      mapping:
+    #        value_for_a: "#/components/schemas/ModelA"
+    #        value_for_b: ModelB   # equivalent to "#/components/schemas/ModelB"
+    #
+    # For any type that isn't specified in the mapping (or if the whole mapping is omitted)
+    # the default lookup value for each schema is the same as the schema name. So this--
+    #      mapping:
+    #        value_for_a: "#/components/schemas/ModelA"
+    # --is exactly equivalent to this:
+    #    discriminator:
+    #      propertyName: modelType
+    #      mapping:
+    #        value_for_a: "#/components/schemas/ModelA"
+    #        ModelB: "#/components/schemas/ModelB"
+
+    def _get_model_name(model: ModelProperty) -> str | None:
+        return get_reference_simple_name(model.ref_path) if model.ref_path else None
+
+    model_types_by_name: dict[str, ModelProperty] = {}
+    for model in subtypes:
+        # Note, model here can never be a UnionProperty, because we've already done
+        # flatten_union_properties() before this point.
+        if not isinstance(model, ModelProperty):
+            return PropertyError(
+                detail="All schema variants must be objects when using a discriminator",
+            )
+        name = _get_model_name(model)
+        if not name:
+            return PropertyError(
+                detail="Inline schema declarations are not allowed when using a discriminator",
+            )
+        model_types_by_name[name] = model
+
+    mapping: dict[str, ModelProperty] = OrderedDict()  # use ordered dict for test determinacy
+    unspecified_models = list(model_types_by_name.values())
+    if data.mapping:
+        for discriminator_value, model_ref in data.mapping.items():
+            if "/" in model_ref:
+                ref_path = parse_reference_path(model_ref)
```

Comment:
As far as I know, there is no other area of OpenAPI/JSON Schema where you can reference a schema by its simple name instead of its path. It's not a general thing, and there is no syntactic way to disambiguate them except by checking for a slash.